language_service/src/lib.rs

```diff
@@ -121,27 +162,81 @@ impl<'a> LanguageService<'a> {
     /// for the document, typically when the document is first opened in the editor.
     /// It should also be called whenever the source code is updated.
     ///
+    /// This is the "entry point" for the language service's logic, after its constructor.
+    ///
     /// LSP: textDocument/didOpen, textDocument/didChange
-    pub fn update_document(&mut self, uri: &str, version: u32, text: &str) {
+    #[async_recursion(?Send)]
+    pub async fn update_document(&mut self, uri: &str, version: u32, text: &str) {
+        if self.currently_updating {
```

Comment:
I can't leave a comment near `document_op`, but you'll want to check this flag within `document_op` and return a default result if we're currently compiling. See: https://github.com/microsoft/qsharp/commit/89e2248316dea51756d353d8e47fd79169e6bd37

Alternatively, instead of checking this flag, you could return a default if the document doesn't exist in the compilation map. The former is more strict about our assumptions, the latter is more likely to gloss over bugs and "just work".

```rs
 trace!("{op_name}: uri: {uri}, offset: {offset}");
        let Some(compilation_uri) = &self
            .open_documents
            .get(uri)
            .as_ref()
            .map(|x| x.compilation.clone())
        else {
            trace!("Skipping {op_name} since compilation is in progress");
            return T::default();
        };
```